crates/store/src/db/sql/mod.rs

```diff
@@ -190,6 +193,317 @@ pub fn select_accounts_by_ids(
     Ok(result)
 }
 
+/// Computes account states for the particular block number using the given [Connection].
+///
+/// # Returns
+///
+/// Account states vector, or an error.
+#[allow(dead_code)] // TODO: remove once the function is used
+pub fn compute_old_account_states(
+    conn: &mut Connection,
+    account_ids: &[AccountId],
+    block_number: BlockNumber,
+) -> Result<Vec<Account>> {
+    fn find_account(
+        accounts: &mut BTreeMap<AccountId, AccountRecord>,
+        account_id: AccountId,
+    ) -> Result<OccupiedEntry<AccountId, AccountRecord>> {
+        let Entry::Occupied(found_account) = accounts.entry(account_id) else {
+            return Err(DatabaseError::DataCorrupted(format!(
+                "Account not found in DB: {account_id:x}"
+            )));
+        };
+
+        Ok(found_account)
+    }
```

Comment:
Maybe we should new-type accounts? Though I'm bias against free standing functions, so take this with a big pinch of 🧂 

```rust
#[derive(Default)]
struct Accounts(BTreeMap<AccountId, AccountRecord>);

impl Accounts {
    fn insert(...)

    // or maybe `get_or_err`
    fn try_get(&mut self) -> Result<&mut AccountRecord> {...}
}
```